windows/src/desktop/kmshell/main/OnlineUpdateCheck.pas

```diff
@@ -706,4 +975,426 @@ function TOnlineUpdateSharedData.Params: TOnlineUpdateCheckParams;
+      // an the install process to execute. We leave the registry value as installing
+      // one installing has finished then the registry value to be set to complete
+      // I also need to consider package updates.
+      // Can the setup.exe program write to the registery to it is complete? is so
+      // will using the msi mean this doesn't work?
+
+    end;
+    installing:
+    begin
+      KL.Log('ProcessBackground Install case :[ installing ]');
+      // TODO if we are in state installing we shouldn't be hear either
+      // So exit and let the msi installer continue.
+    end;
+    postinstall:
+    begin
+      KL.Log('ProcessBackground Install case :[ postinstall ]');
+      // TODO Remove cached files. Do any loging updating of files etc and then set back to idle
+      SavePath := IncludeTrailingPathDelimiter(GetFolderPath(CSIDL_COMMON_APPDATA) + SFolder_CachedUpdateFiles);
+      GetFileNamesInDirectory(SavePath, fileNames);
+      // for now we only want the exe all though excute install can
+      // handle msi and msp
+      for fileName in fileNames do
+      begin
+        System.SysUtils.DeleteFile(fileName);
+      end;
+      // Also remove old versioned dlls if necessary. Initial tests indicate the Microsoft Installer
+      // well do this.
+      SetBackgroundState(idle);
+    end;
+
+  end;
+
+end;
+
+
+function TOnlineUpdateCheck.handleEventCheckForUpdates: Boolean;  // I2329
+var
+  UpdateState : TUpdateState;
+  CheckResult : TOnlineUpdateCheckResult;
+  DownloadResult: Boolean;
+  // For removing cache
+  SavePath: String;
+  fileName: String;
+  fileNames: TStringDynArray;
+begin
+// check the registry value
+  UpdateState := CheckBackgroundState;
+  case UpdateState of
+    idle:
+      begin
+      // Do Nothing
+      end;
+    check, download:  // as the moment check and download are basically the same state
+      begin
+        CheckResult := CheckForUpdates;
+        if CheckResult = oucUpdatesAvailable then
+        begin
+          SetBackgroundState(download);
+          // We can transition straight to download
+          DownloadResult := DownloadUpdatesBackground;
+          if DownloadResult then
+            begin
+              SetBackgroundState(pending);
+              // request install
+              if IsKeymanRunning then
+                // can't install just set icon
+                // then handleInstall will kick it off
+              else
+               // start installing "handleInstall"
+               begin
+                SetBackgroundState(installing);
+                if not BackgroundInstall then
+                // TODO // if BackgroundInstall fails then exit and handle event
+                // back to pending ( 3 times ) then after that set reovery.
+                begin
+                  SetBackgroundState(pending);
+                end;
+               end;
+            end;
+        end
+        else
+          SetBackgroundState(idle);
```

Comment:
I have a hard time finding the corresponding `if`, so I'm wondering if it would make sense to swap the `if` and `else` blocks: 
```
if CheckResult != oucUpdatesAvailable then
  SetBackgroundState(idle);
else
  ...
```

(same logic appears already a few times further up)